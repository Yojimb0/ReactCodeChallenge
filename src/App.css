.App{
  
}

.App header {
  background: url('https://cdn.glitch.com/1c5ff4b7-4717-4c22-9baf-c344d613f454%2FScreen%20Shot%202021-02-21%20at%209.15.20%20pm%20(1).webp?v=1613909251423') center center;
  background-size:cover;
  display: flex;
  align-items: center;
  justify-content: center;
  color: white;
  height: calc(600 / 3000 * 100vw);
}

.App main{
  padding: 30px;
  max-width:500px;
  margin:auto;
}