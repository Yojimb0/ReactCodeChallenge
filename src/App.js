import React, { useState, useEffect } from "react";
import './App.css';
import { MovieComponent } from './components/Movie';
import Total from './styled-components/Total';
import Loading from './styled-components/Loading';
import Error from './styled-components/Error';

export const App = () => {
  const [data, setData] = useState([]);
  const [votes, setVotes] = useState([]);
  const [error, setError] = useState(false);
  const [loading, setLoading] = useState(true);
  
  // Initial fetch
  useEffect(() => {
    const fetchData = async () => {
      try{
        const response = await fetch('https://swapi.dev/api/films/');
		// console.log('fetch response:', response);
		if (!response.ok){
			throw response.statusText;
		}
        const json = await response.json();
        setData(json.results.sort((a,b) => new Date(a.release_date) - new Date(b.release_date)));
      }catch(e){
		// console.log(e);
        setError(e);
      }
    };
    fetchData();
  }, []); // empty dependency array so no updates causes this effect to execute again
  
  // Creating our votes array
  useEffect(() => {
     if (data.length) {
       setVotes(data.map(item=>({episode_id:item.episode_id, votes:0})));
     }
  },[data]);
  
  // Both our data arrays are ready, we can remove the loading label
  useEffect(() => {
     if (votes.length) {
       setLoading(false);
     }
  },[votes]);

  const handleVote = (episode_id) => {
    const clonedVotes = [...votes];
    clonedVotes.find(movie=>movie.episode_id === episode_id).votes++;
    setVotes(clonedVotes);
  }

  return (
    <div className="App">
        <header className="App-header">
          <h1>Star Wars ratings</h1>
        </header>
        <main role="list">
          {error
            ? <Error>An error occured - {error}</Error>
            : loading
              ? <Loading>Loading ...</Loading>
              : data.map((movie,index) => (
                <MovieComponent
                  title={movie.title}
                  releaseDate={movie.release_date}
                  episodeId={movie.episode_id}
                  votes={votes.find(item=>item.episode_id === movie.episode_id).votes}
                  voteHandler={(episodeId) => handleVote(episodeId)}
                  key={movie.episode_id}
				  testId={index}>
                </MovieComponent>
              ))
          }
          <Total>
            <dt>Total number of votes</dt>
            <dd data-testid="totalVotes">{votes.reduce((sum, i) => sum + i.votes, 0)}</dd>
          </Total>
        </main>
      </div>
  );
};
