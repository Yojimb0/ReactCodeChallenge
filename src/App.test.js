import React from 'react';
import nock from 'nock';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
 
import {App} from './App';
 
describe('App', () => {
  
  test('renders App component', () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  test('Shows loading label by default', () => {
    render(<App />);
    screen.getByText('Loading ...');
  });

  test('Shows error when 500 from API', async () => {
    const scope = nock('https://swapi.dev')
      .persist().get('/api/films/')
      .reply(500, '', {'Access-Control-Allow-Origin': '*'});
    
    render(<App />);
    await waitFor(() => screen.getByText('An error occured - Internal Server Error'))
    nock.cleanAll()
  });

  test('Shows correct amount of received items', async () => {
    const scope = nock('https://swapi.dev')
      .persist().get('/api/films/')
      .reply(200, {
        "results": [
          {"title": "Mock1","episode_id": 1,"release_date": "2002-05-16"},
          {"title": "Mock2","episode_id": 2,"release_date": "2001-05-16"},
          {"title": "Oldest","episode_id": 3,"release_date": "1995-05-16"},
          {"title": "Mock3","episode_id": 4,"release_date": "2000-05-16"}
        ]
      }, {'Access-Control-Allow-Origin': '*'});
    
    render(<App />);
    await waitFor(() => screen.getByText('Mock1'));
    const items = screen.getAllByRole('listitem');
    expect(items).toHaveLength(4);
    nock.cleanAll()
  });

  test('Shows oldest at the top', async () => {
    const scope = nock('https://swapi.dev')
      .persist().get('/api/films/')
      .reply(200, {
        "results": [
          {"title": "Mock1","episode_id": 1,"release_date": "2002-05-16"},
          {"title": "Mock2","episode_id": 2,"release_date": "2001-05-16"},
          {"title": "Oldest","episode_id": 3,"release_date": "1995-05-16"},
          {"title": "Mock3","episode_id": 4,"release_date": "2000-05-16"}
        ]
      }, {'Access-Control-Allow-Origin': '*'});
    
    render(<App />);
    const element = await screen.findByTestId('title#0');
    expect(element.textContent).toBe('Oldest');
    nock.cleanAll()
  });

  test('Shows correct amount of votes', async () => {
    const scope = nock('https://swapi.dev')
      .persist().get('/api/films/')
      .reply(200, {
        "results": [
          {"title": "Mock1","episode_id": 1,"release_date": "2002-05-16"},
          {"title": "Mock2","episode_id": 2,"release_date": "2001-05-16"},
          {"title": "Oldest","episode_id": 3,"release_date": "1995-05-16"},
          {"title": "Mock3","episode_id": 4,"release_date": "2000-05-16"}
        ]
      }, {'Access-Control-Allow-Origin': '*'});
    
    render(<App />);
    await screen.findByTestId('button#2');

    fireEvent.click(screen.getByTestId('button#2'));
    fireEvent.click(screen.getByTestId('button#2'));
    expect(screen.getByTestId('votes#2').textContent).toBe('2 votes');
    nock.cleanAll()
  });

  test('Shows correct amount of total votes', async () => {
    const scope = nock('https://swapi.dev')
      .persist().get('/api/films/')
      .reply(200, {
        "results": [
          {"title": "Mock1","episode_id": 1,"release_date": "2002-05-16"},
          {"title": "Mock2","episode_id": 2,"release_date": "2001-05-16"},
          {"title": "Oldest","episode_id": 3,"release_date": "1995-05-16"},
          {"title": "Mock3","episode_id": 4,"release_date": "2000-05-16"}
        ]
      }, {'Access-Control-Allow-Origin': '*'});
    
    render(<App />);
    await screen.findByTestId('button#2');

    fireEvent.click(screen.getByTestId('button#1'));
    fireEvent.click(screen.getByTestId('button#0'));
    fireEvent.click(screen.getByTestId('button#0'));
    fireEvent.click(screen.getByTestId('button#3'));
    fireEvent.click(screen.getByTestId('button#2'));
    expect(screen.getByTestId('totalVotes').textContent).toBe('5');
    nock.cleanAll()
  });
  
});
