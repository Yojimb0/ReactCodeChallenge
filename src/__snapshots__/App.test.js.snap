// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App component 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="App-header"
    >
      <h1>
        Star Wars ratings
      </h1>
    </header>
    <main>
      <p
        class="sc-fubCfw jWxqXS"
      >
        Loading ...
      </p>
      <dl
        class="sc-iBPRYJ bnXwkW"
      >
        <dt>
          Total number of votes
        </dt>
        <dd>
          0
        </dd>
      </dl>
    </main>
  </div>
</div>
`;
